flow/raster_cache.cc

```diff
@@ -26,18 +26,30 @@ RasterCacheResult::RasterCacheResult(sk_sp<SkImage> image,
                                      const char* type)
     : image_(std::move(image)), logical_rect_(logical_rect), flow_(type) {}
 
-void RasterCacheResult::draw(SkCanvas& canvas, const SkPaint* paint) const {
+void RasterCacheResult::draw(SkCanvas& canvas,
+                             const SkPaint* paint,
+                             const SkMatrix* matrix) const {
   TRACE_EVENT0("flutter", "RasterCacheResult::draw");
   SkAutoCanvasRestore auto_restore(&canvas, true);
+  auto device_total_matrix = canvas.getTotalMatrix();
   SkIRect bounds =
-      RasterCache::GetDeviceBounds(logical_rect_, canvas.getTotalMatrix());
+      RasterCache::GetDeviceBounds(logical_rect_, device_total_matrix);
   FML_DCHECK(
       std::abs(bounds.size().width() - image_->dimensions().width()) <= 1 &&
       std::abs(bounds.size().height() - image_->dimensions().height()) <= 1);
   canvas.resetMatrix();
   flow_.Step();
-  canvas.drawImage(image_, bounds.fLeft, bounds.fTop, SkSamplingOptions(),
-                   paint);
+  if (matrix == nullptr) {
+    canvas.drawImage(image_, bounds.fLeft, bounds.fTop, SkSamplingOptions(),
+                     paint);
+  } else {
+    FML_DCHECK(matrix->isScaleTranslate());
```

Comment:
yeah, yeah... I know

Reply:
(I think this is fine - if someone misuses this it'll draw something kind of weird, but it'll still draw something, and having a DCHECK here to make sure we don't misuse it in our own test cases of the API we expose is valid)

Reply:
Oh I meant more that I couldn't figure out how to multiply the matrices so I just did mapRect twice 😢

Reply:
How would they misuse this? Why would it draw something kinda weird?

Oh, I see, because the rendering code below doesn't handle an arbitrary matrix. That can be fixed...

Reply:
The problem here is that the code for the untransformed cache draw is not a very straightforward way to accomplish the goals here and it is largely unnecessary except that it has a very small potential to have slight roundoff errors that spook the underlying graphics engine (Skia) from using a straight pixel-for-pixel blit which was very fast on slow hardware into using a (degenerately unnecessary) scaling GPU shader. As a result, they explicitly dumbed down the straight cache blit to a hard-coded identity-transform blit to avoid any potential for round-off mistakes.

But the code you are adding is already intending to invoke a transformed blit operation anyway, so rather than try to code it to match the identity-space clamped blit operation that cache::draws normally use, just draw the cache entry with the appropriate transform.